substrate/runtime/primitives/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+	PartialOrd<Self> + Ord
+> SimpleArithmetic for T {}
+
+pub trait SimpleBitOps:
+	Sized +
+	rstd::ops::BitOr<Self, Output = Self> +
+	rstd::ops::BitAnd<Self, Output = Self>
+{}
+impl<T:
+	Sized +
+	rstd::ops::BitOr<Self, Output = Self> +
+	rstd::ops::BitAnd<Self, Output = Self>
+> SimpleBitOps for T {}
+
+/// Something that can be executed.
+pub trait Executable {
+	fn execute();
+}
+
+impl Executable for () {
+	fn execute() {}
+}
+impl<A: Executable, B: Executable> Executable for (A, B) {
+	fn execute() {
+		A::execute();
+		B::execute();
+	}
+}
+
+/// Something that acts like a `Digest` - it can have `Log`s `push`ed onto it and these `Log`s are
+/// each `Slicable`.
+pub trait Digesty {
+	type Item: Sized;
+	fn push(&mut self, item: Self::Item);
+}
+
+/// Something which fulfills the abstract idea of a Substrate header. It has types for a `Number`,
+/// a `Hash` and a `Digest`. It provides access to an `extrinsics_root`, `state_root` and
+/// `parent_hash`, as well as a `digest` and a block `number`.
+///
+/// You can also create a `new` one from those fields.
+pub trait Headery: Sized + Slicable {
+	type Number: Sized;
+	type Hash: Sized;
+	type Digest: Sized;
+	fn number(&self) -> &Self::Number;
+	fn extrinsics_root(&self) -> &Self::Hash;
+	fn state_root(&self) -> &Self::Hash;
+	fn parent_hash(&self) -> &Self::Hash;
+	fn digest(&self) -> &Self::Digest;
+	fn new(
+		number: Self::Number,
+		extrinsics_root: Self::Hash,
+		state_root: Self::Hash,
+		parent_hash: Self::Hash,
+		digest: Self::Digest
+	) -> Self;
+}
+
+/// Something which fulfills the abstract idea of a Substrate block. It has types for an
+/// `Extrinsic` piece of information as well as a `Header`.
+///
+/// You can get an iterator over each of the `extrinsics` and retrieve the `header`.
+pub trait Blocky {
+	type Extrinsic: Sized;
+	type Header: Headery;
+	fn header(&self) -> &Self::Header;
+	fn extrinsics(&self) -> &[Self::Extrinsic];
+	fn deconstruct(self) -> (Self::Header, Vec<Self::Extrinsic>);
+}
+
+/// A "checkable" piece of information, used by the standard Substrate Executive in order to
+/// check the validity of a piece of extrinsic information, usually by verifying the signature.
+pub trait Checkable: Sized {
+	type CheckedType: Sized;
+	fn check(self) -> Result<Self::CheckedType, Self>;
+}
+
+/// An "executable" piece of information, used by the standard Substrate Executive in order to
+/// enact a piece of extrinsic information by marshalling and dispatching to a named functioon
+/// call.
+///
+/// Also provides information on to whom this information is attributable and an index that allows
+/// each piece of attributable information to be disambiguated.
+pub trait Applyable {
+	type AccountIdType;
+	type IndexType;
+	fn index(&self) -> &Self::IndexType;
+	fn sender(&self) -> &Self::AccountIdType;
+	fn apply(self);
+}
+
+/// Something that can be checked for equality and printed out to a debug channel if bad.
+pub trait CheckEqual {
```

Comment:
seems like it should be a macro and not a trait

Reply:
not sure what you mean.

Reply:
this functionality seems like it could have a blanket implementation for all `PartialEq` types but actually is just a wrapper for a single implementation (with undocumented reliance on `runtime-io` print helpers) for 256-bit hashes.

Reply:
that's because Hashes are specially printable by `runtime_io`. in any case, it was inherited by previous code and i want to keep the scope of this PR minimal since it's 6k lines and countless hours already.